### PR TITLE
Extensions enhancement

### PR DESCRIPTION
**Pull request by atlas0fd00m**
example_gui_extensions.py gets some examples as well.

### CODE REVIEW

**Review by rakuy0 — CHANGES_REQUESTED** [review]
Mostly some compat stuff. At some point would you mind driving me through thisso I can see it work, since this kind of thing is a mess to test?

File: vivisect/extensions/example_gui_extension.py [hunk 1]
```diff
@@ -19,28 +27,101 @@ def __init__(self, vw, vwgui):
...
+
+        # Add a Button and a Text Edit object in a basic VBox layout
         button = QPushButton('My Button!', parent=self)
         textedit = QTextEdit('WOOT! Some text!', parent=self)
         self.setLayout( VBox(button, textedit) )
 
+
+def vprint(vw, s, *args, **kwargs):
+    vw.vprint(s % args)
+    print s % args
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/extensions/example_gui_extension.py:58]: done

File: vivisect/extensions/example_gui_extension.py [hunk 2]
```diff
@@ -19,28 +27,101 @@ def __init__(self, vw, vwgui):
...
+
+
+def ctxMenuHook(vw, va, expr, menu, parent, nav):
+    '''
+    Context Menu handler (adds options as we wish)
+    '''
+    try:
+        print "foobar"
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/extensions/example_gui_extension.py:66]: deleted

File: vivisect/extensions/example_gui_extension.py [hunk 3]
```diff
@@ -19,28 +27,101 @@ def __init__(self, vw, vwgui):
...
+        if va == 0x41414141:
+            menu.addAction('WAT?',   ACT(vw.vprint, "We're at AAAA!"))
+        menu.addAction('bookmark (B)',   ACT(vw.getVivGui().addBookmark, va))
+        menu.addAction('YEEE HAH',   ACT(vw.vprint, "YEE HAH %x  %r %r %r %r" % (va, expr, menu, parent, nav)))
+        menu.addAction('YEEE HAH1',   ACT(vprint, vw, "YEE HAH %x  %r %r %r %r", va, expr, menu, parent, nav))
+
+    except Exception as e:
+        print e
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/extensions/example_gui_extension.py:74]: deleted

File: vivisect/extensions/example_gui_extension.py [hunk 4]
```diff
@@ -19,28 +27,101 @@ def __init__(self, vw, vwgui):
...
+        self.vw = vw
+        self.vwgui = vwgui
+
+    def thing(self):
+        vprint(self.vw, "Blah Blah Blah")
+
+    def printUserInput(self):
+        # ok is whether the "OK" button was pressed, utext is the user text
+        utext, ok = QInputDialog.getText(self.vwgui, 'Enter...', 'User Text')
+        vprint(self.vw, '%r:  %r', ok, utext)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/extensions/example_gui_extension.py:95]: actually no.  this is as intended.

File: vivisect/qt/ctxmenu.py [hunk 5]
```diff
@@ -216,6 +216,11 @@ def buildContextMenu(vw, va=None, expr=None, menu=None, parent=None, nav=None):
         menu.addAction('undefine (U)',   ACT(vw.delLocation, va))
 
     e_q_memcanvas.initMemSendtoMenu(expr, menu)
-    return menu
 
+    # give any extensions a chance to play
+    for exthook in vw._ext_ctxmenu_hooks:
+        logger.warn('exthook: %r', exthook)
+        print exthook, vw, va, expr, menu, parent, nav
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/qt/ctxmenu.py:223]: deleted and logging changed to info

File: vivisect/__init__.py [hunk 6]
```diff
@@ -88,6 +88,8 @@ def __init__(self, **kwargs):
         else:
             self.vivhome = e_config.gethomedir(".viv", makedir=autosave)
         self._viv_gui = None    # If a gui is running, he will put a ref here...
+        self._ext_ctxmenu_hooks = []
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/__init__.py]: a plague on you and your house.

[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/__init__.py]: ;)  that took longer than all the other changes put together, and introduced fails.  it was more indepth than i had time for.  should be better now.

**Review by rakuy0 — APPROVED** [review]
One question and a couple aesthetic things, but nothing that generally mucks with the functioning.

File: vivisect/__init__.py [hunk 7]
```diff
@@ -192,6 +194,49 @@ def getVivGui(self):
...
+        self._ext_ctxmenu_hooks.pop(handler)
+
+    def addExtension(self, name, extmod):
+        '''
+        Add extension module to a list of extensions.
+        This keeps a list of installed extension modules, with the added value
+        of keeping the loaded module in memory.
+        '''
+        if extmod not in self._extensions:
+            self._extensions[name] = extmod
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/__init__.py]: we can later if we want. currently i think no.

**Review by rakuy0 — APPROVED** [review]

**Pull request merged**